### PR TITLE
Integrate Office 365 SMTP Email Server 

### PR DESCRIPTION
**Pull request by sanjacornelius**
## Issue & Reproduction Steps

Ticket: [FOUR-6103](https://processmaker.atlassian.net/browse/FOUR-6103)

As a System Administrator, I want to enable an email server using the Office 365 SMTP so that my platform utilizes my preferred and familiar email service

This PR updates the Settings UI based on the [Email Server PRD](https://processmaker.atlassian.net/wiki/spaces/PM4/pages/2761064449/Email+Server+Options+-+Office365+Modern+Auth+and+Gmail+API) specifications as well as adds a new SMTP Authentication Method option allowing users to select either Standard, Google or Office 365 Authentication.

## Solution

- Adds the 'office365' option in the UI for the 'SMTP Authentication Method' Model
- Implements XOAUTH2 for Office 365 SMTP

## UI Updates

**Before**

Screen Shot 2022-07-14 at 11 07 28 AM
Screen Shot 2022-07-14 at 11 07 39 AM

**After**

Screen Shot 2022-08-10 at 9 09 35 AM

Screen Shot 2022-08-10 at 9 09 15 AM

## How to Test

**Testing UI and Migration**

1. Review the [PRD](https://processmaker.atlassian.net/wiki/spaces/PM4/pages/2761064449/Email+Server+Options+-+Office365+Modern+Auth+and+Gmail+API)
2. Ensure you are on the `develop` branch for both Core and Send Email Connector
3. Ensure the following packages are installed - 
    - Auth Package
    - Send Email Connector  `develop`
    - Actions By Email
    - DocuSign Connector
4. Configure a few Email servers with and without SMTP settings
5. Update Core branch `feature/FOUR-6103B` 
6. Update Send Email Connector branch `feature/FOUR-6103B`
7. Reinstall Packages
4. Run `php artisan migrate`
8. Within the settings tabs ensure UI displays properly as specified in the PRD
   - Buttons are displayed properly
   - Existing Email Servers Configurations display properly
   - Add / Remove Email Servers works as expected
   - Configurations Update as expected. etc...

## Related Tickets & Packages
- Ticket: [FOUR-6103](https://processmaker.atlassian.net/browse/FOUR-6103)
- Ticket: [FOUR-6104](https://processmaker.atlassian.net/browse/FOUR-6104)
- [Email Server PRD](https://processmaker.atlassian.net/wiki/spaces/PM4/pages/2761064449/Email+Server+Options+-+Office365+Modern+Auth+and+Gmail+API)
- [Send Email PR - Google Integration](https://github.com/ProcessMaker/connector-send-email/pull/211) `feature/FOUR-6104B`

ci:connector-send-email:feature/FOUR-6103B

## Code Review Checklist
- [ ] I have pulled this code locally and tested it on my instance, along with any associated packages.
- [ ] This code adheres to [ProcessMaker Coding Guidelines](https://github.com/ProcessMaker/processmaker/wiki/Coding-Guidelines).
- [ ] This code includes a unit test or an E2E test that tests its functionality, or is covered by an existing test.
- [ ] This solution fixes the bug reported in the original ticket.
- [ ] This solution does not alter the expected output of a component in a way that would break existing Processes.
- [ ] This solution does not implement any breaking changes that would invalidate documentation or cause existing Processes to fail.
- [ ] This solution has been tested with enterprise packages that rely on its functionality and does not introduce bugs in those packages.
- [ ] This code does not duplicate functionality that already exists in the framework or in ProcessMaker.
- [ ] This ticket conforms to the PRD associated with this part of ProcessMaker.

### CODE REVIEW

**Review by nolanpro — CHANGES_REQUESTED** [review]
Here's my suggested changes for both this branch and the package:
- In composer.json, move the Google and Microsoft dependencies to core since they are used there too. 
- Remove require_once  __DIR__ . '/../../vendor/autoload.php'; in EmailController.php controller. It's not needed and often composer install is not run in the package folder.
- Authorize account button should be hidden on standard auth?
- Consistency issues (package and core) with EMAIL_CONNECTOR_OFFICE_* environment variables/settings. Some use an underscore before the index and some don't. We should probably always use the underscore before the index to stay consistent.
- OauthTransportManager:152 is setting $expirationTime to a Carbon object. Although this correctly serializes when stored, it causes scripts to fail because script tasks run escapeshellarg() on it.
- In config/app.php, providers has both MailServiceProvider and OauthMailServiceProvider. OauthMailServiceProvider extends MailServiceProvider and calls its parent registerSwiftTransport(). OauthMailServiceProvider sets a singleton for swift.transport and so does the parent. Since the singleton is overwritten, you can remove the call to parent:: in OauthMailServiceProvider and remove Illuminate\Mail\MailServiceProvider::class from app.php

**Comment by sanjacornelius** [comment]
@nolanpro  Updates have been made to both PRs. Can you please pull down the latest and retest it?

**Review by nolanpro — APPROVED** [review]

File: ProcessMaker/Managers/OauthTransportManager.php [hunk 1]
```diff
@@ -0,0 +1,178 @@
...
+            $expireTime = $now->add(new DateInterval('PT' . $newToken->expires_in . 'S'));
+            $expireDate = $expireTime->format('Y-m-d H:i:s');
+
+            $this->updateEnvVar("EMAIL_CONNECTOR_OFFICE_365_ACCESS_TOKEN{$index}", $newToken->access_token);
+            $this->updateEnvVar("EMAIL_CONNECTOR_OFFICE_365_REFRESH_TOKEN{$index}", $newToken->refresh_token);
+            $this->updateEnvVar("EMAIL_CONNECTOR_OFFICE_365_ACCESS_TOKEN_EXPIRE_DATE{$index}", $expireDate);
+        return $newToken->access_token;
+
+        } catch (Throwable $error) {
+            throw $error;
```
Comment (danloa) [inline ProcessMaker/Managers/OauthTransportManager.php:162]: as the error is simply re-thrown after the catch,  the try/catch block is unnecesary. Add some error handling (log info, etc.) to keep the try/catch

File: ProcessMaker/Managers/OauthTransportManager.php [hunk 2]
```diff
@@ -0,0 +1,178 @@
...
+                    break;
+                case 'office365':
+                    $accessToken = $this->checkForExpiredOffice365AccessToken($this->emailServerIndex);
+                    // Update Authencation Mode
+                    $transport->setAuthMode('XOAUTH2')
+                    ->setUsername($this->fromAddress)
+                    ->setPassword($accessToken);
+                    break;
+                
+                default:
```
Comment (danloa) [inline ProcessMaker/Managers/OauthTransportManager.php:86]: remove the default as there isn't code in it

**Pull request merged**